### DISCUSSION TITLE
How to use the root file system made by myself under the qemu platform?

### DISCUSSION THREAD

**Discussion by Sunt-git** [bao-project/bao-hypervisor#56] [General]
How does a virtual machine running Linux use a self-made root file system instead of a memory file system?

**Comment by josecm** [comment]
@Sunt-git I'm not sure what you mean by "self-made root file system". Do you mean using a "typical" non-volatile storage-based file system instead of the ramfs we use in bao-demos?

I have never done it in qemu. But I'd assume you could just instantiate a QEMU's virtio-block device and pass it through to the Linux guest. [This configuration in bao-demos](https://github.com/bao-project/bao-demos/blob/master/demos/linux%2Bfreertos/configs/qemu-aarch64-virt.c) already assigns all instantiated virtio devices to the linux guest. It also configures the linux guest's memory as identity mapped because these devices are not subject to smmu translation.

Let me know if you were successful with this. Otherwise, I'd be happy to have a crack at it myself when I get some spare cycles.

**Reply by Sunt-git** [reply]
Yes, I mean using a "typical" non-volatile storage-based file system. I have successfully tested this approach using the following QEMU command line options:
`
-drive file=rootfs.ext4,format=raw,id=hd0,if=none \
-device virtio-blk-device,drive=hd0
`
I also added "root=/dev/vda rw" to the Linux kernel startup parameters.
And then, it worked normally.

**Reply by josecm** [reply]
That's great! I'll close the discussion then.
